onnx/defs/math/defs.cc

```diff
@@ -3392,4 +3392,676 @@ ONNX_OPERATOR_SET_SCHEMA(
+                "The signal_dim at the specified axis is equal to the dft_length.",
+                "T1")
+            .TypeConstraint(
+                "T1",
+                {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
+                "Constrain input and output types to float tensors.")
+            .TypeConstraint(
+                "T2",
+                {"tensor(int64)"},
+                "Constrain scalar length types to int64_t.")
+            .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+                bool is_onesided = static_cast<bool>(getAttribute(ctx, "onesided", 0));
+                bool inverse = static_cast<bool>(getAttribute(ctx, "inverse", 0));
+
+                if (inverse && is_onesided) {
+                    fail_shape_inference("is_onesided and inverse attributes cannot be enabled at the same time");
+                }
+
+                propagateElemTypeFromInputToOutput(ctx, 0, 0);
+                if (!hasInputShape(ctx, 0))
+                {
+                    // If no shape is available for the input, skip shape inference...
+                    return;
+                }
+
+                // In general the output shape will match the input shape exactly
+                // So initialize the output shape with the input shape
+                auto& input_shape = getInputShape(ctx, 0);
+                ONNX_NAMESPACE::TensorShapeProto result_shape_proto = input_shape;
+
+                // Get the axis where the DFT will be performed.
+                auto axis = static_cast<int>(getAttribute(ctx, "axis", 1));
+                auto rank = input_shape.dim_size();
+
+                if (!(-rank <= axis && axis < rank)) {
+                    fail_shape_inference(
+                        "axis attribute value ",
+                        axis,
+                        " is invalid for a tensor of rank ",
+                        rank);
+                }
+                
+                auto axis_idx = (axis >= 0 ? axis : axis + rank);
+                
+                // If dft_length is specified, then we should honor the shape.
+                // Set the output dimension to match the dft_length on the axis.
+                // If onesided this will be adjusted later on...
+                const TensorProto* dft_length = nullptr;
+                if (hasInputShape(ctx, 1))
+                {
+                    // If there is a dft_length specified, then we should
+                    // honor it.
+                    if (ctx.getNumInputs() >= 2)
+                    {
+                        dft_length = ctx.getInputData(1);
+                    } else {
+                        // If we cannot read the dft_length, we cannot infer shape
+                        // return...
+                        return;
+                    }
+                }
+
+                if (nullptr != dft_length) {
```

Comment:
I think that 
(a) this if-statement can be moved into the then-branch above, after line 3531.
(b) we should add an else-branch to this condition, clearing value and param (like lines 3561/62)
(c) we can drop the else-branch above in line 3532 (because in that case, we use the input-dim as-is).
Do the changes make sense?

Reply:
Since these suggestions correlate to the previous comment, maybe it is better if I describe the issues I think could be addressed (regardless of how the logic is structured). 
(a) The `return` in like 3535 could be made more precise. We cannot infer the dimension for the fft-axis, but the other dimensions can be propagated as is, correct?

Reply:
I updated the previous check.
I left the return in there because it seemed weird to set a partially specified tensor shape.